data/lm/generate_lm.py

```diff
@@ -1,63 +1,135 @@
+import argparse
 import gzip
 import io
 import os
 import subprocess
-import tempfile
-
 from collections import Counter
 from urllib import request
 
-def main():
-  # Grab corpus.
-  url = 'http://www.openslr.org/resources/11/librispeech-lm-norm.txt.gz'
+import tqdm
+
 
-  with tempfile.TemporaryDirectory() as tmp:
-    data_upper = os.path.join(tmp, 'upper.txt.gz')
-    print('Downloading {} into {}...'.format(url, data_upper))
-    request.urlretrieve(url, data_upper)
+# ======================================================================================================================
+
+def convert_and_filter_topk(args):
+    """ Convert to lowercase, count word occurrences and save top-k words to a file """
 
-    # Convert to lowercase and count word occurences.
     counter = Counter()
-    data_lower = os.path.join(tmp, 'lower.txt.gz')
-    print('Converting to lower case and counting word frequencies...')
-    with io.TextIOWrapper(io.BufferedWriter(gzip.open(data_lower, 'w')), encoding='utf-8') as lower:
-      with io.TextIOWrapper(io.BufferedReader(gzip.open(data_upper)), encoding='utf8') as upper:
-        for line in upper:
-          line_lower = line.lower()
-          counter.update(line_lower.split())
-          lower.write(line_lower)
-
-    # Build pruned LM.
-    lm_path = os.path.join(tmp, 'lm.arpa')
-    print('Creating ARPA file...')
+    data_lower = os.path.join(args.output_dir, 'lower.txt.gz')
+
+    # Convert and count words
+    print('\nConverting to lowercase and counting word occurrences ...')
+    with io.TextIOWrapper(io.BufferedWriter(gzip.open(data_lower, 'w+')), encoding='utf-8') as lower:
+        with open(args.input_txt, encoding='utf8') as upper:
```

Comment:
This is weird, you open `lower.txt.gz`, decode it and then write that to `args.input_txt` which is documented to be the source text. I think it's not what you intended.

Reply:
?
```
with io.TextIOWrapper(io.BufferedWriter(gzip.open(data_lower, 'w+')), encoding='utf-8') as lower:
        with open(args.input_txt, encoding='utf8') as upper:
            for line in tqdm.tqdm(upper):
                line_lower = line.lower()
                counter.update(line_lower.split())
                lower.write(line_lower)
```

I'm opening `args.input_txt` as file `upper` and writing every line into `lower.txt.gz`

But I think I will rename those files for better readability when I'm solving the other suggestions you made.

Reply:
Ah, right, well this is confusing, as you can see. I don't understand why you write to `.txt.gz`

Reply:
This is because i am calling `lmplz ... --text data_lower`. It was called with `.txt.gz` before so I assumed i also need a `.txt.gz` file.